resolver/resolver.cc

```diff
@@ -553,46 +553,69 @@ class ResolveConstantsWalk {
+                }
+                return;
             }
-            return;
-        }
-        if (id->symbol == owner) {
-            if (auto e =
-                    gs.beginError(core::Loc(todo.file, send->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
-                e.setHeader("Must not pass your self to `{}`", send->fun.show(gs));
+            if (id->symbol.data(gs)->isClassOrModuleClass()) {
+                if (auto e =
+                        gs.beginError(core::Loc(todo.file, id->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
+                    e.setHeader("`{}` is a class, not a module; Only modules may be mixins",
+                                id->symbol.data(gs)->show(gs));
+                }
+                encounteredError = true;
             }
-            return;
-        }
-        auto existing = owner.data(gs)->findMember(gs, core::Names::classMethods());
-        if (existing.exists() && existing != id->symbol) {
-            if (auto e =
-                    gs.beginError(core::Loc(todo.file, send->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
-                e.setHeader("Redeclaring `{}` from module `{}` to module `{}`", send->fun.show(gs),
-                            existing.data(gs)->show(gs), id->symbol.data(gs)->show(gs));
+            if (id->symbol == owner) {
+                if (auto e =
+                        gs.beginError(core::Loc(todo.file, id->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
+                    e.setHeader("Must not pass your self to `{}`", send->fun.show(gs));
+                }
+                encounteredError = true;
             }
-            return;
+            if (encounteredError) {
+                continue;
+            }
+
+            // Get the fake property holding the mixes
+            auto mixMethod = owner.data(gs)->findMember(gs, core::Names::mixedInClassMethods());
+            auto loc = core::Loc(owner.data(gs)->loc().file(), send->loc);
+            if (!mixMethod.exists()) {
+                // We never stored a mixin in this symbol
+                // Create a the fake property that will hold the mixed in modules
+                mixMethod = gs.enterMethodSymbol(loc, owner, core::Names::mixedInClassMethods());
+                vector<core::TypePtr> targs;
+                mixMethod.data(gs)->resultType = core::TupleType::build(gs, move(targs));
+
+                // Create a dummy block argument to satisfy sanitycheck during GlobalState::expandNames
+                auto &arg = gs.enterMethodArgumentSymbol(core::Loc::none(), mixMethod, core::Names::blkArg());
+                arg.flags.isBlock = true;
+            }
+
+            auto type = core::make_type<core::ClassType>(id->symbol);
+            (core::cast_type<core::TupleType>(mixMethod.data(gs)->resultType))->elems.emplace_back(type);
```

Comment:
When adding to this vector, could you make sure to only add if it isn't already a member? When sorbet is run in LSP mode, this pass will be re-run for files that have been processed already, so unconditionally appending will cause the vector to grow unbounded. Given that the vectors are expected to be small, it's be fine to just scan through the whole thing and append when you don't find the symbol you're adding.